### ISSUE TITLE
Crash of the program when using ObjectResolver

### ISSUE THREAD

**Issue by ValentinChareyre** [JoshClose/CsvHelper#1148]
Hello,

Following our discussion on the [StackOverflow](https://stackoverflow.com/questions/52970244/using-specific-function-to-create-new-instance-of-object-when-deserializing-csv/52994016), I tried to use the `ObjectResolver` as @JoshClose suggested. However, for a reason I can't explain, my program crashes when I use it. No log, no error, the program crashes.

For your information, I use the CsvHelper library inside the [Unity3D game engine](http://unity3d.com), which uses Mono.

First of all, I have an abstract class responsible for serializing and deserializing data using the library

```csharp
public abstract class DataSerializer<T> : ScriptableObject where T : ScriptableObject
{
    protected abstract CsvHelper.Configuration.ClassMap<T> CSVClassMap { get; }
    protected abstract CsvHelper.IObjectResolver CSVObjectResolver { get; }

    public T[] Deserialize( string data )
    {
        T[] output = DeserializeFromCSV( data );
        
        // Do some stuff
        // Removing it does not prevent the crash

        return output;
    }

    private T[] DeserializeFromCSV( string csv )
    {
        T[] output = null;
        using ( System.IO.TextReader textReader = new System.IO.StringReader( csv ) )
        using ( CsvHelper.CsvReader csvReader = new CsvHelper.CsvReader( textReader ) )
        {
            // Saving current object resolver before assigning a new one
            // Removing the "save" of the object resolver does not prevent the program from crashing
            CsvHelper.IObjectResolver currentObjectResolver = CsvHelper.ObjectResolver.Current;
            CsvHelper.ObjectResolver.Current = CSVObjectResolver;

            // Registering class map to select the needed attributes
            csvReader.Configuration.RegisterClassMap( CSVClassMap );

            // Reading values
            csvReader.Read();
            output = csvReader.GetRecords<T>().ToArray();

            // Restoring object resolver
            CsvHelper.ObjectResolver.Current = currentObjectResolver;
        }
        
        return output;
    }
}
```
----

Then, I have concrete classes implementing the abstract getters, and a CSV class map

```csharp
protected override CsvHelper.Configuration.ClassMap<Ingredient> CSVClassMap
{
    get { return csvClassMap ?? ( csvClassMap = new MyCSVClassMap() ); }
}

protected override CsvHelper.IObjectResolver CSVObjectResolver
{
    get
    {
        // I use the static functions of the CSVClassMap 
        return csvObjectResolver ?? ( csvObjectResolver = new CsvHelper.ObjectResolver( MyCSVClassMap.CanResolve, MyCSVClassMap.Resolve ) );
    }
}

public sealed class MyCSVClassMap : CsvHelper.Configuration.ClassMap<MyClass>
{
    public MyCSVClassMap()
    {
        AutoMap();
        // Ignoring some attributes...
    }

    public static bool CanResolve( Type type )
    {
        return type == typeof( MyClass );
    }
    
    public static object Resolve( Type type, params object[] constructorArgs )
    {
        return ScriptableObject.CreateInstance<MyClass>(); // Using a static method used to instantiate MyClass
    }
}
```
----

Did I do somethign wrong ? Does the `ObjectResolver` part runs on a separate thread? I ask this because I can't call the Unity API outside of the main thread....

Thanks in advance for your help.

**Comment by JoshClose** [comment]
Why are you restoring the object resolver? If things are threaded in your environment, this could cause a problem. You should just set it and leave it. 

The object resolver is meant to be replaced with an IoC container of some sort. It looks like `ScriptableObject.CreateInstance` serves that purpose.

**Comment by ValentinChareyre** [comment]
I restore the Object Resolver to keep the things in the same state I found it. I've tried to remove this part of the code (save + restore), but I got the same result. I don't think a simple `try-catch` will help me, but I will give it a try.

Calling `ScriptableObject.CreateInstance` instead of using the `new` keyword is mandatory because the Unity engine runs additional functions.

**Comment by JoshClose** [comment]
Oh good. I'll look into what might be causing that.

**Comment by JoshClose** [comment]
What type is it trying to resolve when the over happens?

**Comment by ValentinChareyre** [comment]
Here is the class :

    public enum FoodGroup
    {
        /* ... */
    }
    public class Ingredient : ScriptableObject, System.IComparable<Ingredient>, System.IEquatable<Ingredient>
    {
        private int id;
        private new string name;
        private FoodGroup foodGroup;
        private string spriteName;

        public int ID
        {
            get { return id; }
            private set { id = value; }
        }

        public string Name
        {
            get { return name; }
            private set { name = value; }
        }
        public FoodGroup FoodGroup
        {
            get { return foodGroup; }
            private set { foodGroup = value; }
        }
        public string SpriteName
        {
            get
            {                
                return spriteName;
            }
            private set
            {
                // Some stuff
                // Commenting the lines does not prevent the problem from happening
                spriteName = value;

            }
        }

        // My functions ...
    }

**Comment by ValentinChareyre** [comment]
@JoshClose, it seems the problem comes from my data structure. I have tried with an empty class inheriting from `ScriptableObject` (containing no member) and it works fine. I will look deeper into it. Thanks for your time.

**Comment by JoshClose** [comment]
Can you try setting `useFallback` to `false`?

```cs
new CsvHelper.ObjectResolver( SimpleCSVClassMap.CanResolve, SimpleCSVClassMap.Resolve, false )
```

If it throws an exception with the type that can't be resolved, then I think I know what the issue is.

**Comment by JoshClose** [comment]
That's the issue. I was able to reproduce the stack overflow and have a fix for it.

**Comment by JoshClose** [comment]
The fix is in version 8.0.0 on nuget.

**Comment by ValentinChareyre** [comment]
Thanks a lot @JoshClose, I will try it out ASAP !
